sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/MultiValueStringOperatorConversions.java

```diff
@@ -458,6 +578,28 @@ boolean isAllowList()
     }
   }
 
+
+  private static List<DruidExpression> harmonizeNullsMvdArg0OperandList(List<DruidExpression> druidExpressions)
+  {
+    final List<DruidExpression> newArgs;
+    if (druidExpressions.get(0).isDirectColumnAccess()) {
```

Comment:
Why not always harmonize the first arg if it's of type `VARCHAR` or `STRING`? So something like `MV_OVERLAP(LOOKUP(mvd, 'lookupTable'), ARRAY['stuff', 'things'])` would still have the first arg harmonized.

Reply:
hmm, good question. I guess I was trying to avoid weird stuff happening in cases where someone was using one of the other MV_ functions which do not coerce arguments like this like `MV_APPEND` and such. For scalar function arguments like this, we actually _do_ automatically do this harmonization in the native layer because we do it whenever we need to 'apply' a function across an MVD, of which this is a case of that happening.

However, I guess that does create a difference in behavior if there is a mix segments which have single value strings and multi-value strings, since the single value strings would not need applied on the native layer... I hate mvds so so much :sob:

Side note, I hope we can get the higher order function stuff done, because it makes me sad everytime i see a query like this when it really should be `MV_OVERLAP(MAP(x -> LOOKUP(x, 'lookupTable'), mvd), ARRAY['stuff', 'things'])`. Granted, that the native layer will automatically apply this transformation (and when it maps like this it does automatically do this harmonization), but i still find it confusing to look at, and if it _was_ written like this, then I think that it would be strange to harmonize the arg0 of `MV_OVERLAP` instead of harmonizing the `mvd` variable, e.g. `MV_OVERLAP(MAP(x -> LOOKUP(x, 'lookupTable'), MV_HAROMONIZE_NULLS(mvd)), ARRAY['stuff', 'things']).

So... I guess we can apply it all the time, i'll need to update some other tests, especially some of the tests i have with array columns which are wrapping the array columns with `ARRAY_TO_MV` and then testing the `MV_` functions, who do correctly distinguish between `null`, `[]` and `[null]` but now would no longer at least when using with these functions.

Reply:
> So... I guess we can apply it all the time, i'll need to update some other tests, especially some of the tests i have with array columns which are wrapping the array columns with `ARRAY_TO_MV` and then testing the `MV_` functions, who do correctly distinguish between `null`, `[]` and `[null]` but now would no longer at least when using with these functions.

How do you define "correctly" in this sentence? I would think that if you apply `ARRAY_TO_MV`, it's more correct to harmonize the arrays than to not harmonize them. That makes them behave more like they would if they had been ingested as MVDs, which I think is the point of `ARRAY_TO_MV`?

Reply:
> I guess I was trying to avoid weird stuff happening in cases where someone was using one of the other MV_ functions which do not coerce arguments like this like `MV_APPEND` and such.

Do you have an example in mind of the weird stuff?

Reply:
>Do you have an example in mind of the weird stuff?

Well, so today `MV_CONCAT(mvd, ARRAY['foo'])` plans to `array_concat(mvd, array('foo'))` which returns `null` instead of `[null, 'foo']` if the `mvd` row is `null` because we don't coerce those values into `[null]`.

Functions like this are why we turned off the blanket coercion we were doing in the first place (`druid.expressions.homogenizeNullMultiValueStringArrays`) because (at the time at least) it felt too strange for the latter answer to be the behavior, and part of the reasoning for this is because when you do a scan query, which is the only way you can see MVDs as their entire (barring grouping on them with MV_TO_ARRAY) you never see the value as `[null]`, rather it is `null` because it also uses `getObject`.

So if this was the argument and then rewritten into `MV_CONTAINS(MV_HAROMONIZE_NULLS(MV_CONCAT(mvd, ['foo'])), 'foo')` at the native expression layer it would be like `array_contains([null], 'foo')` which is false, instead of `array_contains(null, 'foo')` which is `null`.

>How do you define "correctly" in this sentence? I would think that if you apply ARRAY_TO_MV to a legit array, it's more correct to harmonize the arrays than to not harmonize them. That makes them behave more like they would if they had been ingested as MVDs, which I think is the point of ARRAY_TO_MV?

I guess there is no such thing as "correctly" because mvds themselves are completely inconsistent. I imagined `ARRAY_TO_MV` to primarily be used to get stuff like automatic unnest, and didn't really imagine reducing the fidelity of the underlying arrays to have the same deficiencies as mvds, but I suppose that is another way of thinking about it and be used with native stuff. It gets really weird doing `ARRAY_TO_MV` and then using `MV_` functions that plan to native expressions (which themselves _actually_ turn the mvds into ARRAY<STRING> so that the expressions can handle them), so it didn't occur to me that we would want to do coercion because at the native layer `ARRAY_TO_MV` is basically `cast(array as ARRAY<STRING>)`.